libflux/src/semantic/parser/mod.rs

```diff
@@ -0,0 +1,1740 @@
+            _ => Err("Not a valid basic type")
+        }
+    }
+
+    // parse_array parses an array monotype
+    fn parse_array(& mut self, token: &Token) -> Result<MonoType, &'static str> {
+
+        if token.token_type != TokenType::LEFTSQUAREBRAC {
+            Err("Not a valid array monotype")
+        } else {
+            let mut token = self.next(); 
+
+            // recursively parse the array's monotype
+            let monotype = self.parse_monotype();
+            match monotype {
+                Ok(monotype) => {
+                    token = self.next(); 
+                    if token.token_type == TokenType::RIGHTSQUAREBRAC {
+                        Ok(MonoType::Arr(Box::new(Array(monotype))))
+                    } else {
+                        Err("Array monotype must have right square bracket")
+                    }
+                }
+                Err(e) => Err(e)
+            }
+        }
+    }
+
+    // parse_function parses a single function monotype 
+    fn parse_function(& mut self, token: &Token) -> Result<MonoType, &'static str> {
+        if token.token_type != TokenType::LEFTPAREN {
+            return Err("Function must start with a left paren")
+        }
+        self.next(); 
+        let mut token = self.next(); 
+
+        let mut req_args = HashMap::new(); 
+        let mut opt_args = HashMap::new(); 
+        let mut pipe_arg = None; 
+        
+        loop {
+            if token.token_type == TokenType::IDENTIFIER {
+                if let Ok(arg) = self.parse_required_optional(&token) {
+                    req_args.insert(arg.0, arg.1);
+                } else {
+                    return Err("Must have valid required arguments")
+                }
+            } else if token.token_type == TokenType::QUESTIONMARK {
+                token = self.next(); // skip question mark 
+                
+                // now we can parse this optional argument the same way 
+                // that we parse required arguments 
+                if let Ok(arg) = self.parse_required_optional(&token){
+                    opt_args.insert(arg.0, arg.1);
+                } else {
+                    return Err("Invalid format for optional arguments")
+                } 
+            } else if token.token_type == TokenType::PIPE {
+                let arg = self.parse_pipe(); 
+                if arg.is_none() {
+                    return Err("Invalid format for pipe arguments")
+                } else {
+                    pipe_arg = arg; 
+                }
+            } else {
+                return Err("Invalid arguments for this function.")
+            }
+            
+            token = self.next(); // check if next is right paren or comma
+            if token.token_type != TokenType::COMMA {
+                // if the next token is not comma, must be right paren
+                break;
+            } 
+            token = self.next(); // if its a comma, then there are more args to parse
```

Comment:
@jsternberg you suggested that I try and use a match statement in here. Was that just a suggestion to make this a little more idiomatic or was there another reason for that change? I'm not seeing the added value in this particular case.